stl/inc/span

```diff
@@ -0,0 +1,758 @@
+    }
+
+    // [span.iterators] Iterator support
+    _NODISCARD constexpr iterator begin() noexcept {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+        return iterator(_Mydata, _Mydata, _Mydata + size());
+#else // ^^^ _ITERATOR_DEBUG_LEVEL >= 1 ^^^ // vvv _ITERATOR_DEBUG_LEVEL == 0 vvv
+        return iterator(_Mydata);
+#endif // _ITERATOR_DEBUG_LEVEL
+    }
+
+    _NODISCARD constexpr iterator end() noexcept {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+        return iterator(_Mydata + size(), _Mydata, _Mydata + size());
+#else // ^^^ _ITERATOR_DEBUG_LEVEL >= 1 ^^^ // vvv _ITERATOR_DEBUG_LEVEL == 0 vvv
+        return iterator(_Mydata + size());
+#endif // _ITERATOR_DEBUG_LEVEL
+    }
+
+    _NODISCARD constexpr const_iterator begin() const noexcept {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+        return const_iterator(_Mydata, _Mydata, _Mydata + size());
+#else // ^^^ _ITERATOR_DEBUG_LEVEL >= 1 ^^^ // vvv _ITERATOR_DEBUG_LEVEL == 0 vvv
+        return const_iterator(_Mydata);
+#endif // _ITERATOR_DEBUG_LEVEL
+    }
+
+    _NODISCARD constexpr const_iterator end() const noexcept {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+        return const_iterator(_Mydata + size(), _Mydata, _Mydata + size());
+#else // ^^^ _ITERATOR_DEBUG_LEVEL >= 1 ^^^ // vvv _ITERATOR_DEBUG_LEVEL == 0 vvv
+        return const_iterator(_Mydata + size());
+#endif // _ITERATOR_DEBUG_LEVEL
+    }
+
+    _NODISCARD constexpr const_iterator cbegin() const noexcept {
+        return begin();
+    }
+
+    _NODISCARD constexpr const_iterator cend() const noexcept {
+        return end();
+    }
+
+    _NODISCARD constexpr reverse_iterator rbegin() noexcept {
+        return reverse_iterator{end()};
+    }
+
+    _NODISCARD constexpr reverse_iterator rend() noexcept {
+        return reverse_iterator{begin()};
+    }
+
+    _NODISCARD constexpr const_reverse_iterator rbegin() const noexcept {
+        return const_reverse_iterator{end()};
+    }
+
+    _NODISCARD constexpr const_reverse_iterator rend() const noexcept {
+        return const_reverse_iterator{begin()};
+    }
+
+    _NODISCARD constexpr const_reverse_iterator crbegin() const noexcept {
+        return rbegin();
+    }
+
+    _NODISCARD constexpr const_reverse_iterator crend() const noexcept {
+        return rend();
+    }
+
+    _NODISCARD constexpr pointer _Unchecked_begin() noexcept {
+        return _Mydata;
+    }
+
+    _NODISCARD constexpr const_pointer _Unchecked_begin() const noexcept {
+        return _Mydata;
+    }
+
+    _NODISCARD constexpr pointer _Unchecked_end() noexcept {
+        return _Mydata + size();
+    }
+
+    _NODISCARD constexpr const_pointer _Unchecked_end() const noexcept {
+        return _Mydata + size();
+    }
+
+private:
+    pointer _Mydata{nullptr};
+};
+
+// [span.tuple] Tuple interface
+template <class _Ty, size_t _Extent>
+struct tuple_size<span<_Ty, _Extent>> : integral_constant<size_t, _Extent> {
+    static_assert(dynamic_extent != _Extent, "std::span<T, dynamic_extent> is not tuple-like");

```

Comment:
In `N4842` this is marked as `not defined` for `dynamic_extend`. I assume that in this case we need to SFINAE out rather than static assert?

Reply:
Good catch. I believe that the proper thing to do is to implement a partial specialization for `span<_Ty, dynamic_extent>` exactly as depicted in the header synopsis (declared but not defined), instead of attempting to perform SFINAE in the `span<_Ty, _Extent>` partial specialization.

Reply:
Done